hal/src/sercom/async_dma.rs

```diff
@@ -0,0 +1,155 @@
+//! Use the DMA Controller to perform async transfers using the SERCOM
+//! peripheral
+//!
+//! See the [`mod@uart`], [`mod@i2c`] and [`mod@spi`] modules for the
+//! corresponding DMA transfer implementations.
+
+use atsamd_hal_macros::hal_macro_helper;
+
+use crate::dmac::{AnyChannel, Beat, Buffer, Error, ReadyFuture, TriggerAction};
+use core::ops::Range;
+
+use super::Sercom;
+
+/// Wrapper type over an `&[T]` that can be used as a source buffer for DMA
+/// transfers. This is an implementation detail to make async SERCOM-DMA
+/// transfers work. Should not be used outside of this crate.
+///
+/// # Safety
+///
+/// [`ImmutableSlice`]s should only ever be used as **source** buffers for DMA
+/// transfers, and never as destination buffers.
+#[doc(hidden)]
+pub(crate) struct ImmutableSlice<T: Beat>(Range<*mut T>);
+
+impl<T: Beat> ImmutableSlice<T> {
+    #[inline]
+    pub(in super::super) fn from_slice(slice: &[T]) -> Self {
+        let ptrs = slice.as_ptr_range();
+
+        let ptrs = Range {
+            start: ptrs.start.cast_mut(),
+            end: ptrs.end.cast_mut(),
+        };
+
+        ImmutableSlice(ptrs)
+    }
+}
+
+unsafe impl<T: Beat> Buffer for ImmutableSlice<T> {
+    type Beat = T;
+    #[inline]
+    fn dma_ptr(&mut self) -> *mut Self::Beat {
+        if self.incrementing() {
+            self.0.end
+        } else {
+            self.0.start
+        }
+    }
+
+    #[inline]
+    fn incrementing(&self) -> bool {
+        self.buffer_len() > 1
+    }
+
+    #[inline]
+    fn buffer_len(&self) -> usize {
+        self.0.end as usize - self.0.start as usize
+    }
+}
+
+/// Wrapper type over Sercom instances to get around lifetime issues when using
+/// one as a DMA source/destination buffer. This is an implementation detail to
+/// make async SERCOM-DMA transfers work.
+#[doc(hidden)]
+#[derive(Clone)]
+pub(crate) struct SercomPtr<T: Beat>(pub(in super::super) *mut T);
+
+unsafe impl<T: Beat> Buffer for SercomPtr<T> {
+    type Beat = T;
+
+    #[inline]
+    fn dma_ptr(&mut self) -> *mut Self::Beat {
+        self.0
+    }
+
+    #[inline]
+    fn incrementing(&self) -> bool {
+        false
+    }
+
+    #[inline]
+    fn buffer_len(&self) -> usize {
+        1
```

Comment:
why 1?

Reply:
So the SERCOM peripherals expect all data to be read/written to the same register address. In the case of a transmission, it then may add it to an internal buffer, after which the data is transmitted (the reverse happens for receive transactions). Therefore the DMA transfer must direct all the incoming/outgoing data to the same memory address, regardless of the transaction length.